### ISSUE TITLE
esm and cjs config file support

### ISSUE THREAD

**Issue by xiaoxiangmoe** [prettier/prettier#7049]
use `prettier.config.cjs` and `prettier.config.mjs` for config file.

**Comment by lydell** [comment]
Hi! Can you link to docs about these file endings? Are other tools supporting this? What are the benefits?

**Comment by evilebottnawi** [comment]
@lydell we should open issue https://github.com/davidtheclark/cosmiconfig, I think support for these extensions should be implemented not in prettier, other tools will probably want to have the same support in the future

**Comment by SMotaal** [comment]
Glad to see `.cjs` and `.mjs` out here.

This comes from Node.js finally unflagging ESM (ie `--experimental-modules`).

The [Node.js Modules team](https://github.com/nodejs/modules/) took time to explore matters of source disambiguation where unlike browsers that only ever deal with `text/javascript` files based on clear intent of loading, tooling and Node.js itself cannot always assume the format of `.js` files.

Package authors can indicate that `.js` is either ESM of CommonJS using the `package.type` field — and so to close the gap for tools expecting CommonJS configuration a `package.type = "module"` (ie ESM) the `.cjs` extension may be necessary when relying on `require(configuration)` in their implementation.

Hope this helps.

**Comment by akauppi** [comment]
Situation in Dec 2020 (prettier 2.2.1):

One can use a `.cjs` configuration file in `package.type = "module"` project. Plain `.js` does not work which is pretty common still these days.

**Beware of this:** (could deserve to be a separate issue, but no time, sorry..)

If there is a syntax error in the JavaScript configuration file (`prettier.config.cjs`), `prettier` seems to ignore the file and try looking for something else. My error message was:

```
$ npm run _format.js
...
> prettier --config ./prettier.config.cjs --write ./src/**/*.js ./app/**/*.js

[error] Invalid configuration file `src/Draggable/draggableX.js`: Unexpected identifier
npm ERR! code 2
npm ERR! path /Users/.../Git/svelte-with-SVG
npm ERR! command failed
npm ERR! command sh -c prettier --config ./prettier.config.cjs --write ./src/**/*.js ./app/**/*.js

npm ERR! A complete log of this run can be found in:
npm ERR!     /Users/.../.npm/_logs/2020-12-16T10_14_00_410Z-debug.log
```

I would expect to rather get an error stating that the config file `prettier.config.cjs` is bad.

**Comment by akauppi** [comment]
As to @lydell 's question (Nov 2019) on the benefits:

- one can use comments in JavaScript config, unlike in JSON
- if one works on fully ES module projects, it makes sense to use the same `import`/`export` abstraction also for the config files. If this is not possible, it is great to be able to mark the deviant files by their suffix: `.cjs`

I think this is becoming the norm but as @alexander-akait wrote, the support should come from elsewhere - no need to hardcode the logic in each node package. Unfortunately, I don't have more the details on that.

**Comment by xHyroM** [comment]
progress?

**Comment by fisker** [comment]
Fixed by #8890 and #13130
